concurrency/worker/priority.go

```diff
@@ -0,0 +1,176 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gostdlib/base/concurrency/worker/internal/heap"
+)
+
+// QJob represents a job to be done in a priority queue.
+type QJob struct {
+	// Priority is the job's priority.
+	Priority uint64
+	// Work is the work to be done by the job.
+	Work func()
+
+	// submit is the submit the job was submitted.
+	submit time.Time
+}
+
+// queue implements the heap interface. We are using a custom generic heap instead of the stdlib.
+type queue struct {
+	mu   sync.Mutex
+	jobs []QJob
+}
+
+func (p *queue) Len() int {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return len(p.jobs)
+}
+
+func (p *queue) Less(i, j int) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Make submission time a tiebreaker.
+	if p.jobs[i].Priority == p.jobs[j].Priority {
+		return p.jobs[i].submit.Before(p.jobs[j].submit)
+	}
+
+	return p.jobs[i].Priority > p.jobs[j].Priority
+}
+
+func (p *queue) Swap(i, j int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.jobs[i], p.jobs[j] = p.jobs[j], p.jobs[i]
+}
+
+func (p *queue) Push(ctx context.Context, x QJob) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.jobs = append(p.jobs, x)
+}
+
+func (p *queue) Pop(ctx context.Context) QJob {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if len(p.jobs) == 0 {
+		return QJob{}
+	}
+
+	n := len(p.jobs) - 1
+	job := p.jobs[n]
+	p.jobs = p.jobs[0:n]
+	return job
+}
+
+// Queue represents a priority queue for jobs. This can be created from a Limited Pool via Queue().
+// If two jobs have the same priority, the job that was submitted first will be processed first.
+type Queue struct {
+	queue *queue
+	done  chan struct{}
+	count atomic.Int64
+	wait  sync.WaitGroup
+	size  chan struct{}
+	pool  *Limited
+}
+
+// Close closes the queue. Be sure that the queue is empty before closing.
+func (d *Queue) Close() {
+	close(d.done)
+}
+
+// QueueLen returns the size of the queue not processed. This does not include QJobs that are
+// currently being processed.
+func (d *Queue) QueueLen() int {
+	return int(d.count.Load())
+}
+
+// Wait waits for the queue to be empty and not processing being done or the context to be canceled.
+// If the context is canceled, the context error will be returned.
+func (d *Queue) Wait(ctx context.Context) error {
+	done := make(chan struct{})
+
+	Default().Submit(
+		ctx,
+		func() {
+			for {
+				if d.QueueLen() == 0 {
```

Comment:
Is the only way the queue loses items through `Pop`? If so, you could have it send an event via channel (or whatever other mechanism) rather than doing a poll here?

Reply:
Ok maybe that's not the right way to do it, as Put/Pop should just be the heap operations possibly - but I think in `doWork` you could write to a channel that this read, rather than checking here, when count == 0?

Reply:
Maybe it doesn't matter because this is in `Wait` anyway and the expectation is we won't be blocking forever

Reply:
For the most part, this is probably okay as it is.   But I went ahead and put a channel mechanism in so that when Pop() is called on queue{} that is shoves a struct{} into a channel if the length of the queue is 0, unless that channel is full.  

I don't know if this is more efficient or not really.  It is a little harder to grok, but that's probably fine.